<div style="border:solid black 2px; padding: 20px">

**Final Project Number _Two_**
    
</div>

<div style="border:solid black 2px; padding: 20px">
<div class="alert alert-block alert-info">

# Books Sparkers Apps Analysis for Startup Company - by awrb
    
</div>

# Table of Contents: <a id='tableofcontents'></a>

    - 1. Introduction
        - 1.1 Goal
        - 1.2 Stages
    - 2. Data Pre-processing
    - 3. Data Study
    - 4. Explorative Data Analysis (EDA)
        - 3.1 Computing the number of books released after January 1, 2000
        - 3.2 Computing the number of user reviews and the average rating for each book
        - 3.3 Identifying the publisher that has published the most number of books, with more than 50 pages
        - 3.4 Identifying authors with the highest average book rating
        - 3.5 Computing the average number of review texts among users who have rated more than 50 books
    - 5. Presentation
    - 6. General Conclusions and Recommendations

## Introduction

        The corona virus whose presence shocked the whole world has changed everyone's daily routine. Now, city residents no longer spend their free time outside the home, such as going to cafes and malls. They are more often at home, spending time reading books. This also encourages startup companies to develop new applications for book sparkers.
    
        From developing this new application, startup companies could open their horizons to the silence of the pandemic and open various opportunities that are offered to all users for a life that continues to grow together.

### Goal

        We have been got a database from one of the competing companies in this industry. The database contains data about books, publishers, authors, as well as customer ratings and reviews of related books. We will use this information in making a price quote for a new product.
        
        And we will elaborate and display those discoveries in this project.

### Stages

    There are several steps to complete this project, namely:

    Step 1. Accessing Database and Loading The Data
    Step 2. Studying The Tables
    Step 3. Performing SQL Queries for Several Tasks
    Step 4. Presenting My Project to The Clients
    Step 5. Formulating General Conclusions and Recommendations for Working with Clients

    which is divided into the following detailed stages:

    1. Data Pre-processing
    2. Data Study
    3. Explorative Data Analysis (EDA)
    4. Presentation
    5. General Conclusions and Recommendations

    With the following description:

    1. Data Pre-processing (Step 1. Accessing Database and Loading The Data)
    - In this stage, we will access and load the Database by using the following database code files path:
    

    import pandas as pd
    from sqlalchemy import create_engine

    db_config = {'user': 'practicum_student',         
                 'pwd': 's65BlTKV3faNIGhmvJVzOqhs', 
                 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
                 'port': 6432,              
                 'db': 'data-analyst-final-project-db'}          

    connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                         db_config['pwd'],
                                                                           db_config['host'],
                                                                           db_config['port'],
                                                                           db_config['db'])

    engine = create_engine(connection_string, connect_args={'sslmode':'require'})
    
    
    And the connection is stored in the 'engine' variable. We can run SQL queries using Pandas:
    
    pd.io.sql.read_sql(query, con = engine)
    
    
    But before we load the codes above, we should download the files of 'https://storage.yandexcloud.net/cloud-certs/CA.pem' to the loading folder 'file .ipynb' (which the use is when working with the database).
   
    Here are the data descriptions of the database code files path above:

    1) 'books' tables contains the books datas which consisting of:
    a. 'book_id' — book ID
    b. 'author_id' — author ID 
    c. 'title' — title of the books
    d. 'num_pages' — number of pages
    e. 'publication_date' — publication date (the date the books were published)
    f. 'publisher_id' — publisher ID

    2) 'authors' tables contains the authors datas which consisting of:
    a. 'author_id' — author ID
    b. 'author' — authors' name

    3) 'publishers' tables contains the publishers datas which consisting of:
    a. 'publisher_id' — publisher ID
    b. 'publisher' — publishers' name

    4) 'ratings' tables contains the user reviews datas which consisting of:
    a. 'rating_id' — rating ID
    b. 'book_id' — book ID
    c. 'username' — the username that rated the books
    d. 'rating' — standardized assessment by user reviews (in scale 1-5 points)

    5) 'reviews' tables contains the customers reviews datas which consisting of:
    a. 'review_id' — review ID
    b. 'book_id' — book ID
    c. 'username' — the name of the user who reviewed the books
    d. 'text' — review text
    
    
    2. Data Study (Step 2. Studying The Tables)
    - In this stage, we will check and ensure for the 'books' Tables, 'authors' Tables, 'publishers' Tables, 'ratings' Tables, and 'reviews' Tables. We will study the tables coherently and thorough.

    
    3. Explorative Data Analysis (EDA) (Step 3. Performing SQL Queries for Several Tasks)
    - In this stage, there are several things that need to be our focus, namely:

    1. Computing the number of books released after January 1, 2000
    2. Computing the number of user reviews and the average rating for each book
    3. Identifying the publisher that has published the most number of books, with more than 50 pages
        - (This will assist us in excluding brochures and similar publications from the analysis).
    4. Identifying authors with the highest average book rating
        - (Finding books with the rating of at least 50).
    5. Computing the average number of review texts among users who have rated more than 50 books
    
    
    4. Presentation (Step 4. Presenting My Project to The Clients)
    - In this stage, I will present My Project to The Clients.
    
    
    5. General Conclusions and Recommendations (Step 4. Formulating General Conclusions and Recommendations for Working with Customers)
    - In this stage, we will draw the conclusions and formulate the recommendations.

[Back to Contents](#tableofcontents)

# Books Sparkers Apps Analysis

## Data Pre-processing
    - Step 1. Accessing Database and Loading The Data

<div style="border:solid black 2px; padding: 20px">    
    
    First of all, we are going to access the Database. Then, load the required libraries, 'pandas', 'numpy', 'create_engine', as follows:
    
</div>  

In [2]:
import pandas as pd
import numpy as np
from sqlalchemy import create_engine

db_config = {'user': 'practicum_student',         
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', 
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              
             'db': 'data-analyst-final-project-db'}          

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

    We have already access and load the Database above. Now, we are going to continue to the Data Study below:

[Back to Contents](#tableofcontents)

## Data Study
    - Step 2. Studying The Tables

<div style="border:solid black 2px; padding: 20px">
    
    First, we will check and ensure for the 'books' tables. In this stage, we will study the tables coherently and thorough.
    
</div>    

# 'books' Tables

In [3]:
pd.io.sql.read_sql(
'''
SELECT
    *
FROM 
    books
''',
con = engine)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

    Yes, we did it. We have successfully displayed the datas above.

    Beforehand, we have to ensure for the missing values in the books tables that already displayed, as follows:

In [4]:
pd.io.sql.read_sql(
'''
SELECT
    *
FROM
    books
WHERE
    books ISNULL
''',
con = engine)

Empty DataFrame
Columns: [book_id, author_id, title, num_pages, publication_date, publisher_id]
Index: []

    For far, the data looks good and we could continue the process.
    Next, we will get the first and the last publication date of books, as follows:

In [5]:
pd.io.sql.read_sql(
'''
SELECT
    MIN (publication_date),
    MAX (publication_date)
FROM
    books
''',
con = engine)

min         max
0  1952-12-01  2020-03-31

    Temporary Findings:
    - The first books publication date is 1952-12-01.
    - The last books publication date is 2020-03-31.

<div style="border:solid black 2px; padding: 20px">

    In this step, we will dig deeper for the initial lines which those are:
    
    1. 'authors' Tables
    2. 'publishers' Tables
    3. 'ratings' Tables
    4. 'reviews' Tables
    
</div>    

    We are going to start in displaying the several initial lines, as follows:

[Back to Contents](#tableofcontents)

# 1. 'authors' Tables

    First, we will check the 'authors' Tables, as follows:

In [6]:
pd.io.sql.read_sql(
'''
SELECT
    *
FROM
    authors
''',
con = engine)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

    Then, we ensure the missing values in the 'authors' Tables, as follows:

In [7]:
pd.io.sql.read_sql(
'''
SELECT
    *
FROM
    authors
WHERE
    authors ISNULL
''',
con = engine)

Empty DataFrame
Columns: [author_id, author]
Index: []

[Back to Contents](#tableofcontents)

# 2. 'publishers' Tables

    Second, we will check the 'publishers' Tables, as follows:

In [8]:
pd.io.sql.read_sql(
'''
SELECT 
    *
FROM 
    publishers
''',
con = engine)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

    Then, we ensure the missing values in the 'publishers' Tables, as follows:

In [10]:
pd.io.sql.read_sql(
'''
SELECT *
FROM publishers
WHERE publishers ISNULL
''',
con = engine)

Empty DataFrame
Columns: [publisher_id, publisher]
Index: []

[Back to Contents](#tableofcontents)

# 3. 'ratings' Tables

    Third, we will check the 'ratings' Tables, as follows:

In [10]:
pd.io.sql.read_sql(
'''
SELECT
    *
FROM
    ratings
''',
con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

    Then, we ensure the missing values in the 'ratings' Tables, as follows:

In [11]:
pd.io.sql.read_sql(
'''
SELECT *
FROM ratings
WHERE ratings ISNULL
''',
con = engine)

Empty DataFrame
Columns: [rating_id, book_id, username, rating]
Index: []

[Back to Contents](#tableofcontents)

# 4. 'reviews' Tables

    Last, we will check the 'reviews' Tables, as follows:

In [12]:
pd.io.sql.read_sql(
'''
SELECT
    *
FROM
    reviews
''',
con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

    Then, we ensure the missing values in the 'reviews' Tables, as follows:

In [13]:
pd.io.sql.read_sql(
'''
SELECT
    *
FROM
    reviews
WHERE
    reviews ISNULL
''',
con = engine)

Empty DataFrame
Columns: [review_id, book_id, username, text]
Index: []

    Findings:
    
    - There are five tables in database, include the 'books' table. Those are:
        1. 'books' Tables
        2. 'authors' Tables
        3. 'publishers' Tables
        4. 'ratings' Tables
        5. 'reviews' Tables

    With the following description:
    
    1. 'books' contains the books datas which consisting of:
    a. 'book_id' — book ID
    b. 'author_id' — author ID 
    c. 'title' — title of the books
    d. 'num_pages' — number of pages
    e. 'publication_date' — publication date (the date the books were published)
    f. 'publisher_id' — publisher ID

    2. 'authors' contains the authors datas which consisting of:
    a. 'author_id' — author ID
    b. 'author' — authors' name

    3. 'publishers' contains the publishers datas which consisting of:
    a. 'publisher_id' — publisher ID
    b. 'publisher' — publishers' name

    4. 'ratings' contains the user reviews datas which consisting of:
    a. 'rating_id' — rating ID
    b. 'book_id' — book ID
    c. 'username' — the username that rated the books
    d. 'rating' — standardized assessment by user reviews (in scale 1-5 points)

    5. 'reviews' contains the customers reviews datas which consisting of:
    a. 'review_id' — review ID
    b. 'book_id' — book ID
    c. 'username' — the name of the user who reviewed the books
    d. 'text' — review text
        
    - The first books publication date is 1952-12-01.
    - The last books publication date is 2020-03-31.
    - There are no missing values in each tables.

[Back to Contents](#tableofcontents)

## Explorative Data Analysis (EDA)
    - Step 3. Performing SQL Queries for Several Tasks

<div style="border:solid black 2px; padding: 20px">
    
    In this stage, there are several things that need to be our focus, namely:
    
    1. Computing the number of books released after January 1, 2000
    2. Computing the number of user reviews and the average rating for each book
    3. Identifying the publisher that has published the most number of books, with more than 50 pages
    4. Identifying authors with the highest average book rating
    5. Computing the average number of review texts among users who have rated more than 50 books
    
</div>

### Computing the number of books released after January 1, 2000

    First, we will compute the number of books released after January 1, 2000, as follows:

In [15]:
pd.io.sql.read_sql(
'''
SELECT
    COUNT(title)
FROM 
    books
WHERE 
    publication_date > '2000-01-01'
''', 
con = engine)

count
0    819

    Findings:
    
    - According to the result above, the number of books released after January 1, 2000 are 819 books.

[Back to Contents](#tableofcontents)

### Computing the number of user reviews and the average rating for each book
    In this session, for computing the number of user reviews and the average rating for each book, we should some steps below:

    1. We are going to use CTE (Common Table Expression) for creating two new tables of:
        a. Getting the reviews number of each book that we will create as table number one.
        b. Getting the ratings average of each book that we will create as table number two.

    2. Both of new tables are going to merge by using INNER JOIN method at 'book_id'.
    
    The execution of computing number of user reviews and average rating for each book, as follows:

    1. CTE (Common Table Expression)

In [19]:
pd.io.sql.read_sql(
'''
WITH agg_reviews_per_book AS (
    SELECT
        reviews.book_id,
        COUNT(reviews.text) AS cnt_reviews,
        books.title
    FROM
        reviews
        INNER JOIN books ON books.book_id = reviews.book_id
    GROUP BY
        reviews.book_id,
        books.title
    ORDER BY
        cnt_reviews DESC
)
, avg_ratings_per_book AS(
    SELECT
        ratings.book_id,
        AVG(ratings.rating) AS avg_ratings
    FROM 
        ratings
    GROUP BY
        ratings.book_id
    ORDER BY
        avg_ratings DESC
)
SELECT 
    agg_reviews_per_book.title,
    cnt_reviews,
    avg_ratings
FROM
    agg_reviews_per_book
    INNER JOIN avg_ratings_per_book on agg_reviews_per_book.book_id = avg_ratings_per_book.book_id
''',
con = engine)

title  cnt_reviews  \
0                                         'Salem's Lot            2   
1                   1 000 Places to See Before You Die            1   
2    13 Little Blue Envelopes (Little Blue Envelope...            3   
3    1491: New Revelations of the Americas Before C...            2   
4                                                 1776            4   
..                                                 ...          ...   
989           Wyrd Sisters (Discworld  #6; Witches #2)            3   
990                        Xenocide (Ender's Saga  #3)            3   
991                                    Year of Wonders            4   
992                        You Suck (A Love Story  #2)            2   
993  Zen and the Art of Motorcycle Maintenance: An ...            4   

     avg_ratings  
0       3.666667  
1       2.500000  
2       4.666667  
3       4.500000  
4       4.000000  
..           ...  
989     3.666667  
990     3.400000  
991     3.200000  
992     4.500000  
993     3.833333  

[994 rows x 3 columns]

    2. INNER JOIN method at 'book_id'

In [20]:
pd.io.sql.read_sql(
'''
WITH main AS (
    WITH agg_reviews_per_book AS (
        SELECT
            reviews.book_id,
            COUNT(reviews.text) AS cnt_reviews,
            books.title
        FROM
            reviews
            INNER JOIN books ON books.book_id = reviews.book_id
        GROUP BY
            reviews.book_id,
            books.title
        ORDER BY
            cnt_reviews DESC
    )
    , avg_ratings_per_book AS(
        SELECT
            ratings.book_id,
            AVG(ratings.rating) AS avg_ratings
        FROM 
            ratings
        GROUP BY
            ratings.book_id
        ORDER BY
            avg_ratings DESC
    )
    SELECT 
        agg_reviews_per_book.title,
        cnt_reviews,
        avg_ratings
    FROM
        agg_reviews_per_book
        INNER JOIN avg_ratings_per_book on agg_reviews_per_book.book_id = avg_ratings_per_book.book_id
    )
SELECT
    MIN(cnt_reviews) AS min_reviews_per_book,
    MAX(cnt_reviews) AS max_reviews_per_book,
    MIN(avg_ratings) AS min_ratings_per_book,
    MAX(avg_ratings) AS max_ratings_per_book
FROM
    main
''',
con = engine)

min_reviews_per_book  max_reviews_per_book  min_ratings_per_book  \
0                     1                     7                   1.5   

   max_ratings_per_book  
0                   5.0

    Findings:
    
    - From the tables above, the books that receive reviews from users are 994 books.
    - The range of reviews per book is 1 of 7 reviews.
    - The range of ratings average is from scale 1.5 to 5.0.

[Back to Contents](#tableofcontents)

### Identifying the publisher that has published the most number of books, with more than 50 pages
    - This will assist us in excluding brochures and similar publications from the analysis.

    We are going to identify the publisher that has published the most number of books, with more than 50 pages, as follows:

In [21]:
pd.io.sql.read_sql(
'''
SELECT 
    publisher,
    COUNT(publisher) AS cnt
FROM 
    publishers
    LEFT JOIN books ON books.publisher_id = publishers.publisher_id
WHERE
    books.num_pages > 50
GROUP BY
    publisher
ORDER BY
    cnt DESC;
''',
con = engine)

publisher  cnt
0                                        Penguin Books   42
1                                              Vintage   31
2                             Grand Central Publishing   25
3                                     Penguin Classics   24
4                                     Ballantine Books   19
..                                                 ...  ...
329                                         Turtleback    1
330  Atheneum Books for Young Readers: Richard Jack...    1
331                                     Penguin Signet    1
332                                    Victor Gollancz    1
333                      Harvard Business Review Press    1

[334 rows x 2 columns]

    Findings:
    
    - Based on data above, there are 334 publishers that has published the most number of books with more than 50 pages.

[Back to Contents](#tableofcontents)

### Identifying authors with the highest average book rating
    - Finding books with the rating of at least 50

    First, we are going to identify the author with the highest average book rating, as follows:

In [22]:
pd.io.sql.read_sql(
'''
SELECT
    authors.author_id,
    authors.author,
    COUNT(ratings.rating) AS cnt_rating,
    ROUND(AVG(ratings.rating), 2) AS avg_rating
FROM 
    books
    LEFT JOIN authors
        ON books.author_id = authors.author_id
    LEFT JOIN ratings
        ON books.book_id = ratings.book_id
GROUP BY
    authors.author_id,
    authors.author
HAVING
    COUNT(ratings.rating) >= 50
ORDER BY
    avg_rating DESC;
''',
con = engine)

author_id                                             author  cnt_rating  \
0         130                                     Diana Gabaldon          50   
1         236                         J.K. Rowling/Mary GrandPré         312   
2           3                                    Agatha Christie          53   
3         402                  Markus Zusak/Cao Xuân Việt Khương          53   
4         240                                     J.R.R. Tolkien         166   
5         499                           Roald Dahl/Quentin Blake          62   
6         376                                  Louisa May Alcott          54   
7         498                                       Rick Riordan          84   
8          39                                      Arthur Golden          56   
9         542                                       Stephen King         106   
10        302                                       John Grisham          70   
11        621                                    William Golding          71   
12        536                                    Sophie Kinsella          57   
13        291                                       Jodi Picoult          76   
14        445                                    Nicholas Sparks         111   
15        257                                    James Patterson          64   
16        235                                      J.D. Salinger          98   
17        630  William Shakespeare/Paul Werstine/Barbara A. M...          66   
18        469           Paulo Coelho/Alan R. Clarke/Özdemir İnce          57   
19        372                                         Lois Lowry          65   
20        106                                          Dan Brown         143   
21        195            George Orwell/Boris Grabnar/Peter Škerl          74   
22        554                                    Stephenie Meyer         160   
23        311                                     John Steinbeck          73   

    avg_rating  
0         4.30  
1         4.29  
2         4.28  
3         4.26  
4         4.24  
5         4.21  
6         4.20  
7         4.13  
8         4.11  
9         4.01  
10        3.97  
11        3.90  
12        3.88  
13        3.88  
14        3.88  
15        3.86  
16        3.85  
17        3.79  
18        3.79  
19        3.74  
20        3.74  
21        3.73  
22        3.66  
23        3.64

    Findings:
    
    - There are 24 authors with the highest average book rating.

[Back to Contents](#tableofcontents)

### Computing the average number of review texts among users who have rated more than 50 books
    In this session, we will do as the same previously method, those are:

    1. We are going to use CTE (Common Table Expression) for creating two new tables of:
    a. Getting the number of users who rated more than 50 books that we will create as table one.
    b. Getting the number of users who review books that we will create as table two.

    2. Both of new tables are going to merge by using INNER JOIN method at 'username'.

    The execution of computing average number of review texts among users who have rated more than 50 books, as follows:

    CTE (Common Table Expression) and INNER JOIN method at 'username':

In [25]:
pd.io.sql.read_sql(
'''
WITH user_more_than_50_ratings AS (
    SELECT
        username,
        COUNT(username) AS cnt_rating
    FROM 
        ratings
    GROUP BY
        username
    HAVING
        COUNT(username) > 50
)
, main AS (
    SELECT
        reviews.username,
        COUNT(reviews.review_id) AS cnt_reviews
    FROM
        reviews
        INNER JOIN user_more_than_50_ratings ON reviews.username = user_more_than_50_ratings.username
    GROUP BY
        reviews.username
    ORDER BY
        cnt_reviews DESC
)
SELECT
    ROUND(AVG(cnt_reviews), 0) AS avg_reviews
FROM
    main
''',
con = engine)

avg_reviews
0         24.0

    Findings:
    
    - Based on the result above, the average number of review texts among users who have rated more than 50 books is 24.0 (24 reviews).

[Back to Contents](#tableofcontents)

## Presentation

<div style="border:solid black 2px; padding: 20px">
    
    This is my presentation of 'Books Sparkers Apps Analysis for Startup Company' project by me, awrb.

</div>

Presentation Link: https://drive.google.com/file/d/18DN0SlH4IBDZQJZmazA2o2J-oPMdkZMl/view?usp=sharing

[Back to Contents](#tableofcontents)

## General Conclusions and Recommendations
    - Step 4. Formulating General Conclusions and Recommendations for Working with Clients

<div style="border:solid black 2px; padding: 20px">
    
# For Data Pre-processing
    
    We have already access and load the Database by using the following database code file path:
    
    import pandas as pd
    from sqlalchemy import create_engine

    db_config = {'user': 'practicum_student',         
                 'pwd': 's65BlTKV3faNIGhmvJVzOqhs', 
                 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
                 'port': 6432,              
                 'db': 'data-analyst-final-project-db'}          

    connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                         db_config['pwd'],
                                                                           db_config['host'],
                                                                           db_config['port'],
                                                                           db_config['db'])

    engine = create_engine(connection_string, connect_args={'sslmode':'require'})


    And the connection is stored in the 'engine' variable. We can run SQL queries using Pandas:

    pd.io.sql.read_sql(query, con = engine)
    
</div>

<div style="border:solid black 2px; padding: 20px">
    
# For Data Study
    
    Findings:
    - There are five tables in database, include the 'books' table. Those are:
    1. 'books' Tables
    2. 'authors' Tables
    3. 'publishers' Tables
    4. 'ratings' Tables
    5. 'reviews' Tables

    With the following description:

    1. 'books' contains the books datas which consisting of:
        a. 'book_id' — book ID
        b. 'author_id' — author ID 
        c. 'title' — title of the books
        d. 'num_pages' — number of pages
        e. 'publication_date' — publication date (the date the books were published)
        f. 'publisher_id' — publisher ID

    2. 'authors' contains the authors datas which consisting of:
        a. 'author_id' — author ID
        b. 'author' — authors' name

    3. 'publishers' contains the publishers datas which consisting of:
        a. 'publisher_id' — publisher ID
        b. 'publisher' — publishers' name

    4. 'ratings' contains the user reviews datas which consisting of:
        a. 'rating_id' — rating ID
        b. 'book_id' — book ID
        c. 'username' — the username that rated the books
        d. 'rating' — standardized assessment by user reviews (in scale 1-5 points)

    5. 'reviews' contains the customers reviews datas which consisting of:
        a. 'review_id' — review ID
        b. 'book_id' — book ID
        c. 'username' — the name of the user who reviewed the books
        d. 'text' — review text

    - The first books publication date is 1952-12-01.
    - The last books publication date is 2020-03-31.
    - There are no missing values in each tables.

</div>

<div style="border:solid black 2px; padding: 20px">
    
# For Explorative Data Analysis (EDA)
    
    1. Computing the number of books released after January 1, 2000
    Findings:
    - According to the result above, the number of books released after January 1, 2000 are 819 books.
    
    2. Computing the number of user reviews and the average rating for each book
    Findings:
    - From the tables above, the books that receive reviews from users are 994 books.
    - The range of reviews per book is 1 of 7 reviews.
    - The range of ratings average is from scale 1.5 to 5.0.
    
    3. Identifying the publisher that has published the most number of books, with more than 50 pages
    Findings:
    - Based on data above, there are 334 publishers that has published the most number of books with more than 50 pages.
    
    4. Identifying authors with the highest average book rating
    Findings:
    - There are 24 authors with the highest average book rating.
    
    5. Computing the average number of review texts among users who have rated more than 50 books
    Findings:
    - Based on the result above, the average number of review texts among users who have rated more than 50 books is 24.0 (24 reviews).
    
</div>

<div style="border:solid black 2px; padding: 20px">

# Presentation

    Presentation Link: https://drive.google.com/file/d/18DN0SlH4IBDZQJZmazA2o2J-oPMdkZMl/view?usp=sharing
    
</div>

<div style="border:solid black 2px; padding: 20px">
    
# Recommendations
    
    - For increasing the next bookselling, we recommend to select the authors with the highest average ratings. It could predict and mean that the authors' highest average ratings, the highest profit possibility happened.
    
    - We have the sizeable number of 24 reviews for the average number of review texts among users who have rated more than 50 books. From this result, we could conclude that the genres they read are interested them, so we recommend to use the genres users interested for getting high review texts.


</div>

[Back to Contents](#tableofcontents)